### ISSUE TITLE
Model checkpoints and classifier accuracy, training ADF

### ISSUE THREAD

**Issue by jaypatravali** [mattiasegu/uncertainty_estimation_deep_learning#2]
Hi,

Few issues:

1. Just tested out your code to train resnet ADF with and without dropout, it gives NaNs for loss. The pretrained models you provide are only for resnet18 and resnet18 with dropout. Can you provide models for resnetADF with and without dropout.
In eval.py
2. The accuracy for testing all networks [resnet18, resnet18 with MCDO, resnet18ADF, Resnet18ADF with MCDO are almost the same. I am not sure whats missing here.

I plotted accuracy, brier score on all four methods running your code. 
1. acc.png 
![acc](https://user-images.githubusercontent.com/8754443/76379661-f1c27880-630d-11ea-80f9-252f719fc46a.png)

The x-axis is 100% in sample test data (Cifar 10) and then reducing the number of in-distribution test samples by adding more out-of-distribution test samples from CIFAR100. 

Similar for Brier score. 
![brier](https://user-images.githubusercontent.com/8754443/76379760-4239d600-630e-11ea-9e98-0d3e0da5ec31.png)

Brier score reference:http://papers.nips.cc/paper/9547-can-you-trust-your-models-uncertainty-evaluating-predictive-uncertainty-under-dataset-shift.pdf

**Comment by antonilo** [comment]
@mattiasegu please have a look

**Comment by mattiasegu** [comment]
Hi @jaypatravali 

Thanks for the interest in our work.

1. Concerning the NaNs for loss when training ADF, it is a known problem that training with the heteroscedastic loss may be difficult because of numerical instability problems. We also mentioned in our paper that it wasn't possible to train the heteroscedastic neural network from [Kendall et al.](https://arxiv.org/pdf/1703.04977.pdf) because of numerical instability enhanced by the SoftMax layer. To address this problem when trainining the ADF network with the heteroscedastic loss, we initialized the network weights from the best pretrained ckpt on Resnet-18 with and without dropout. You can try it yourself, no modification to the code are needed, you only need to load one of the two available ckpts. Nevertheless, we will update the trained ckpts for ADF as soon as possible.

2. To answer to the second question, I need to know what flags you are using. To replicate the performance reported in the paper for our method, you will need to use `p=0.02, min_variance=1e-3, noise_variance=1e-3`, whereas the network with dropout at training time was trained with `p=0.2` and hence MCDO should be tested with that parameter. Finally, how many Monte-Carlo samples are you collecting? The performance reported in the paper are for 20 samples.

With this command you should be able to replicate the results for the our method (different runs will report slightly different results because of the stochasticity of MCDO):

```
python eval.py -r -b -m \
    --load_model_name resnet18 \
    --test_model_name resnet18_dropout_adf \
    --p 0.02 \
    --min_variance 1e-3 \
    --noise_variance 1e-3 \
    --num_samples 20
```

On a side note, if you only look at the accuracy, you can find values for p that make our method perform better. We found that `p=0.02` was the best tradeoff in terms of accuracy, log-likelihood and Brier score (tested but not reported in the paper for space constraints).

**Comment by DanCunnington** [comment]
Hi @mattiasegu thanks for a very interesting paper and codebase. I'd also be interested in the ADF checkpoints when available please 👍

**Comment by mattiasegu** [comment]
Hi @DanCunnington 

I am glad you liked our work! The ADF checkpoints have just been added to the folder linked in the README, sorry for the delay.

**Comment by DanCunnington** [comment]
Hi @mattiasegu - thank you. I've tried running your code and have a couple of observations..

1. I had to add `map_location=torch.device('cpu')` to the [torch.load function](https://github.com/mattiasegu/uncertainty_estimation_deep_learning/blob/master/eval.py#L123) for running on a CPU.

2. I also had to add `strict=False`to the [load_state_dict function](https://github.com/mattiasegu/uncertainty_estimation_deep_learning/blob/master/eval.py#L125) - which suggests non matching keys. Is the network you used to obtain the checkpoints the same as the network in this repository?

3. Following your instructions in the README, classification performance is very poor and i'm also getting 'nans' for uncertainty values. Running:

```
python eval.py -r -b -m \                                                                                                                      
    --load_model_name resnet18 \
    --test_model_name resnet18_dropout_adf \
    --p 0.02 \
    --min_variance 1e-3 \
    --noise_variance 1e-3 \
    --num_samples 20

==> Preparing data...
Files already downloaded and verified
Files already downloaded and verified
==> Building model...
==> Resuming from checkpoint..
Loaded checkpoint at location ./checkpoint/ckpt_resnet18.pth
==> Loaded model statistics:
        test_model_name = resnet18_dropout_adf
        load_model_name = resnet18
        @epoch = 276
        best_acc    = 94.9
==> Selected parameters:
        use_mcdo       = True
        num_samples    = 20
        p              = 0.02
        min_variance   = 0.001
        noise_variance = 0.001
        tau = 0.0001
==> Starting evaluation...
 [============================ 100/100 ===========================>]  Step: 1m58s | Tot: 3h26m | Loss: 2.303 | Acc: 10.000% (1000/10000)                                                                                                    
Accuracy                = 10.0
Brier Score             = 0.09000000953674317
Negative log-likelihood = 44106.8914
Time                    = 12514.10853099823
```
results in 10% classification accuracy and also `tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],...]...)` when printing the contents of the `outputs_variance` variable in the [evaluate](https://github.com/mattiasegu/uncertainty_estimation_deep_learning/blob/master/eval.py#L205) function.

4. Loading the ADF checkpoint, with:

```
python eval.py -r -b -m \                                                                                                                      
    --load_model_name resnet18_dropout_adf \
    --test_model_name resnet18_dropout_adf \
    --p 0.02 \
    --min_variance 1e-3 \
    --noise_variance 1e-3 \
    --num_samples 20
```

The variance outputs all appear to be 0.001, which is equal to the `min_variance` passed in.

Is there something I'm missing for steps 3 and 4?

Thanks

**Comment by mattiasegu** [comment]
Hi @DanCunnington 

Concerning the observations 1, 2 and 3, I don't meet any of those issues. Could you please tell me what PyTorch version are you using? Step 3 could be a consequence of steps 1 and 2. Did you try running the code in the provided virtual environment?

**Comment by mattiasegu** [comment]
Hi @DanCunnington

1-2-3. Unfortunately, I cannot reproduce any of those issues. Could you please tell me what PyTorch version are you using? Step 3 could be a consequence of steps 1 and 2. Did you try running the code in the provided virtual environment?

4. From my experiments, the variance output is often close to 0.001, but never exactly 0.001. The reason is that, for this very simple task, the model uncertainty is little compared to the data uncertainty, which accounts for most of the total uncertainty. Moreover, you may have noticed that there is another input parameter `--noise_variance`, which represents the variance that we inject at every layer. Since ADF relies on the assumption that all intermediate activations in the network can be approximated by independent gaussians, it underestimates the aleatoric variance. We need `--noise_variance` to compensate for that. As I said before, this happens in this very simple task, where the data availability is large and explains out the model uncertainty. In the other tasks analysed in the paper, where little data were available, the model variance accounted already for most of the total variance.

**Comment by DanCunnington** [comment]
Thanks for your response. I'm using PyTorch version `1.4.0` and running with python `3.7` in a virtual environment:

```
virtualenv -p /usr/bin/python3 .
./bin/activate
pip install -r requirements.txt
python eval.py ... (as above)
```

**Comment by mattiasegu** [comment]
I tried to replicate your issue on another machine, using the same configuration as yours and the checkpoints uploaded online. This is the result:

```
python eval.py -r -b -m \                                                                                                                      
    --load_model_name resnet18 \
    --test_model_name resnet18_dropout_adf \
    --p 0.02 \
    --min_variance 1e-3 \
    --noise_variance 1e-3 \
    --num_samples 20


Files already downloaded and verified
==> Building model...
==> Resuming from checkpoint..
Loaded checkpoint at location ./checkpoint/ckpt_resnet18.pth
==> Loaded model statistics:
        test_model_name = resnet18_dropout_adf
        load_model_name = resnet18
        @epoch = 276
        best_acc    = 94.9
==> Selected parameters:
        use_mcdo       = True
        num_samples    = 20
        p              = 0.02
        min_variance   = 0.001
        noise_variance = 0.001
        tau = 0.0001
==> Starting evaluation...

[================================================================>]  Step: 5s649ms | To 100/100 | Loss: 1.558 | Acc: 94.000% (9400/10000)
Accuracy                = 94.0
Brier Score             = 0.009081093919277191
Negative log-likelihood = 272.6534275407791
Time                    = 560.1841177940369
```

Is it possible that you didn't pull/clone the latest version of the repository? Little modifications were done and checkpoint loading could be affected. Can you tell me if you still have the same problem afterward?

**Comment by mattiasegu** [comment]
Hi @DanCunnington , I see that you are running on CPU, probably loading the model with `strict=False` disregards mismatches in model loading. In a related pull request #4 @marbakken implemented a quick fix that should solve your issue.
